### PR TITLE
Reliable updating of email-based indices

### PR DESCRIPTION
**Pull request by indeyets**
1. drop old index if email has changed
2. do not create index for empty email

### CODE REVIEW

**Comment by berkus** [comment]
Look gut now.

**Comment by berkus** [comment]
Lets not merge it at the moment.

**Comment by indeyets** [comment]
This pull-request needs one more fix:
during the update, we need to be sure, that email-index is not taken. otherwise we can end up with situation, when several users have same email address, but index points only to the last one, meaning that other accounts are not able to "restore" the password

**Comment by indeyets** [comment]
the check should, probably, go to `User.prototype.emailIsValid`

File: test/unit/user.js [hunk 1]
```diff
@@ -44,20 +44,17 @@ describe('User', function() {
   describe('#isValidUsername()', function() {
     var valid = ['luna', '12345', 'hello1234',
                  ' group', 'group '] // automatically trims
+    var i = 1
     valid.forEach(function(username) {
-      it('should allow username ' + username, function(done) {
+      it(`should allow username '${username}'`, function(done) {
```
Comment (epicmonkey) [inline test/unit/user.js:53]: LIEK!

File: app/models/user.js [hunk 2]
```diff
@@ -297,6 +311,8 @@ exports.addModel = function(database) {
 
   User.prototype.update = async function(params) {
     var has_changes = false
+      , email_changed = false
+      , old_email = ""
```
Comment (epicmonkey) [inline app/models/user.js]: Nitpicking: email_changed -> emailChanged, old_email -> oldEmail

File: app/models/user.js [hunk 3]
```diff
@@ -180,18 +181,27 @@ exports.addModel = function(database) {
...
-  User.prototype.isValidEmail = function() {
-    var valid = this.emailIsValid(this.email)
-
-    return Promise.resolve(valid)
+  User.prototype.isValidEmail = async function() {
+    return User.emailIsValid(this.email)
   }
 
-  User.prototype.emailIsValid = function(email) {
+  User.emailIsValid = async function(email) {
```
Comment (epicmonkey) [inline app/models/user.js:188]: Let's extract to Validator module, which does not exist yet. Could be in a separate PR.

**Comment by epicmonkey** [comment]
Need a couple of tests:
- userA, userB are in the system, userA changes email to newEmail, userB can change email to userA
- userA, userB are in the system, userA changes email to newEmail, userB changes email to userA, userA can reset password over newEmail
- userA, userB are in the system, userA changes email to newEmail, userB changes email to userA, userB can reset password over email userA 
- userA, userB are in the system, userB changes email to userA -> error

**Comment by indeyets** [comment]
@epicmonkey I am not sure what would be the proper way to test that proper user's password is reset

**Comment by epicmonkey** [comment]
Let's just make sure you can fire controller's action like here: https://github.com/pepyatka/pepyatka-server/blob/development/test/functional/passwords.js#L32
